Add --json and filtering options for --export-db to udevadm info

DaanDeMeyer opened the pull request:
<!-- devel-freezer = {"comment-id":"1633121595","freezing-tag":"v254-rc1"} -->

DaanDeMeyer commented:
I would really like multiple property matches to behave as if they were joined by AND by default but this is not supported see https://github.com/systemd/systemd/issues/28372

DaanDeMeyer commented:
I'm also wondering whether a new verb is OK for this. It didn't seem easy to implement this as part of `info` though, especially since I'd like to use `Table` for the json output support.

DaanDeMeyer commented:
Reworked to make this part of `udevadm info --export-db`. We don't use `Table` anymore to keep the non json output the same as before. I kept the commits adding glyphs and the delimiter stuff for table since I think it could still be useful later, but they're not required anymore.

DaanDeMeyer commented:
Still need to add docs and such but would be great to make sure this approach is the right one before doing that.

DaanDeMeyer commented:
Updated to use `sd_device_enumerator_add_match_property_required()` so that all given properties are matched against. Still need to write docs and add a basic test.

Updated to use our "streaming" JSON output since there could potentially be a lot of devices.

yuwata commented on src/udev/udevadm-info.c:
```diff
@@ -662,41 +798,96 @@ static int print_tree(sd_device* below) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse key/value pair %s: %m", s);
+        if (r < 2)
+                return log_error_errno(SYNTHETIC_ERRNO(r), "Missing '=' in key/value pair %s.", s);
```
Maybe, better to check that the input string does not contain multiple `=`, e.g. `KEY=VAL=FOO` ?

DaanDeMeyer replied:
Would that be a problem? AFAIK there's nothing preventing the kernel from putting '=' in property values.

poettering replied:
`r` is not a valid error, first arg should be changed to `SYNTHETIC_ERRNO(EINVAL)` or so

yuwata reviewed (changes requested):
Looks nice! Several minor comments.

poettering commented on src/basic/glyph-util.c:
```diff
@@ -132,6 +134,10 @@ const char *special_glyph_full(SpecialGlyph code, bool force_utf) {
+
+                        /* Single glyphs in both cases */
+                        [SPECIAL_GLYPH_LINE_HORIZONTAL]         = u8"⎯",         /* actually called: HORIZONTAL LINE EXTENSION */
+                        [SPECIAL_GLYPH_LINE_VERTICAL]           = u8"｜",         /* actually called: VERTICAL LINE EXTENSION */
```
hmm, this probably should use boxdrawing chars, i.e.:

https://www.unicode.org/charts/PDF/U2500.pdf

poettering commented on src/shared/format-table.c:
```diff
@@ -2522,6 +2544,14 @@ int table_print(Table *t, FILE *f) {
+                                        fputs(special_glyph(SPECIAL_GLYPH_LINE_HORIZONTAL), f);
+
+                        fputc('\n', f);
+                }
```
so the lines separating the unit types in the "systemctl" output are currently drawn via ANSI underline stuff. should we do the same here?

there should probably be only one way how we draw horizontal lines in tables i guess.

DaanDeMeyer replied:
I find the ANSI underline stuff rather ugly. I'd much rather we switch to using this new thing everywhere.

poettering commented on src/shared/format-table.c:
```diff
@@ -156,6 +156,9 @@ struct Table {
 
         bool *reverse_map;
+
+        size_t *delimiter_map; /* In vertical mode, after which rows a delimiter should be added. */
```
why only in vertical mode btw? in regular systemctl outut we also draw horizontal lines, we should probably unify that...

poettering commented on src/shared/format-table.c:
```diff
@@ -2916,40 +2945,43 @@ static int table_to_json_vertical(Table *t, JsonVariant **ret) {
 
-        return json_variant_new_object(ret, elements, n_elements);
+        *ret = a ? TAKE_PTR(a) : TAKE_PTR(o);
+        return 0;
```
hmm, i wonder if outputting an array is really the right approach here. if you have a huge number of elements you stream out, it might be nicer to generate json-seq instead, because it means you can output/parse this stuff piecemeal and not keep the whole array in memory.

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -260,6 +265,117 @@ static int print_record(sd_device *device, const char *prefix) {
+        assert(ret);
+
+        assert_se(sd_device_get_devpath(device, &str) >= 0);
+        r = json_variant_set_field_string(&v, "P", str);
```
uh, why not use strings such as "devpath" for the json output?

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -260,6 +265,117 @@ static int print_record(sd_device *device, const char *prefix) {
+                        return -ENOMEM;
+
+                r = json_variant_set_field_string(&v, "D", str);
+                if (r < 0)
```
if we do json let's do json, i.e. generate a sub object that consists of a string and two chars, i.e. `( "type" : "char", "major" : 4711, "minor" : 1234 )`

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -260,6 +265,117 @@ static int print_record(sd_device *device, const char *prefix) {
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
```
it's ok to use json_variant_set_field_string() if you must, but also kinda in-effecient. Please use json_build() instead, i.e. query all the bits, and then do:

```c
r = json_build(ret, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(diskseq != UINT64, JSON_BUILD_INTEGER(diskseq)),
JSON_BUILD_PAIR_CONDITION(devnum != 0, JSON_BUILD_OBJECT(JSON_BUILD_PAIR("type", JSON_BUILD_STRING(streq(subsys, "block") ? "block" : "char"), JSON_BUILD_PAIR("major", JSON_BUILD_INTEGER(major(devnum)), …
```

you get the idea

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -499,7 +618,24 @@ static int help(void) {
+               "     --name-match=NAME        Query devices with this /dev name\n"
+               "     --parent-match=NAME      Query devices with this parent device\n"
+               "     --initialized-match      Query devices that are already initialized\n"
+               "     --initialized-nomatch    Query devices that are not initialized yet\n",
```
any chance the json stuff can be done in a separate commit from the match stuff?

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -662,41 +798,96 @@ static int print_tree(sd_device* below) {
+                return log_error_errno(SYNTHETIC_ERRNO(r), "Missing '=' in key/value pair %s.", s);
+
+        if (!filename_is_valid(k))
+                return log_error_errno(r, "%s is not a valid key name", k);
```
SYNTHETIC_ERRNO(EINVAL)

DaanDeMeyer commented on src/udev/udevadm-info.c:
```diff
@@ -260,6 +265,117 @@ static int print_record(sd_device *device, const char *prefix) {
+                             DEVNUM_FORMAT_VAL(devnum)) < 0)
+                        return -ENOMEM;
+
+                r = json_variant_set_field_string(&v, "D", str);
```
Yeah let's do this in another PR

DaanDeMeyer commented:
I dropped the glyph and delimiter commits for now, will put those up as another PR at some point. Split the last commit into two, one for --json and one for the filtering.

Also dropped most of the shorthand properties from the JSON output as we already get these as properties.

DaanDeMeyer commented:
Added some tests now as well, PTAL!

bluca commented on src/udev/udevadm-info.c:
```diff
@@ -723,43 +732,96 @@ static int print_tree(sd_device* below) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse key/value pair %s: %m", s);
+        if (r < 2)
+                return log_error_errno(SYNTHETIC_ERRNO(r), "Missing '=' in key/value pair %s.", s);
```
did you mean `SYNTHETIC_ERRNO(EINVAL)` here?

bluca commented on src/udev/udevadm-info.c:
```diff
@@ -723,43 +732,96 @@ static int print_tree(sd_device* below) {
+                return log_error_errno(SYNTHETIC_ERRNO(r), "Missing '=' in key/value pair %s.", s);
+
+        if (!filename_is_valid(k))
+                return log_error_errno(SYNTHETIC_ERRNO(r), "%s is not a valid key name", k);
```
as above

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -723,43 +732,96 @@ static int print_tree(sd_device* below) {
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return r;
```
info_main() logs about all errors, but not these ones, please fix

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -858,6 +920,70 @@ int info_main(int argc, char *argv[], void *userdata) {
+
+                case ARG_ATTR_MATCH:
+                case ARG_ATTR_NOMATCH:
+                        r = parse_key_value_argument(optarg, &k, &v);
```
this leaks memory if used twice on the same cmdline. k/v should be declared in a local `{}` block here, so that they are freed right away, after we used them

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -858,6 +920,70 @@ int info_main(int argc, char *argv[], void *userdata) {
+                        break;
+
+                case ARG_PROPERTY_MATCH:
+                        r = parse_key_value_argument(optarg, &k, &v);
```
ditto

DaanDeMeyer commented:
Minor changes, so setting the green label again

poettering commented on src/udev/udevadm-info.c:
```diff
@@ -873,7 +1044,7 @@ int info_main(int argc, char *argv[], void *userdata) {
 
         if (action == ACTION_EXPORT)
-                return export_devices();
+                return export_devices(e);
```
i guess this needs an ensure_device_enumerator() too
